### PR TITLE
Changed axis order of structure mask, added lecture demos

### PR DESCRIPTION
**Pull request by SimonBiggs**
* [x] Added a clearer contract for the structure mask, it will now raise a ValueError instead of running off and calculating a nonsense mask
* [x] Made the mask axis order be `z, y, x` to match `pymedphys.dicom.zyx_and_dose_from_dataset`
* [x] Added a rudimentary test that covers a simple case

### CODE REVIEW

**Comment by SimonBiggs** [comment]
To view PRs that have a lot of prototyping notebooks in them, and you just want to ignore those, you can click the following:

image

File: pymedphys/_dicom/dose.py [hunk 1]
```diff
@@ -179,58 +179,81 @@ def profile(displacements, depth, direction, dose_dataset, plan_dataset):
...
+    mask_yxz = np.zeros((len(y_dose), len(x_dose), len(z_dose)), dtype=bool)
 
-    mask = np.zeros((len(y_dose), len(x_dose), len(z_dose)), dtype=bool)
+    for structure_index, z_val in enumerate(structure_z_values):
+        dose_index = int(np.where(z_dose == z_val)[0])
 
-    for z_val in structure_z_values:
-        structure_indices = _get_indices(z_structure, z_val)
+        if z_structure[structure_index][0] != z_dose[dose_index]:
+            raise ValueError("Expected structure and dose indices to align")
```
Comment (sjswerdloff) [inline pymedphys/_dicom/dose.py:223]: I think it is clearer if you state what the error is, explicitly.
"Structure and dose indices do not align"

but this isn't a reason to hold up the PR

File: pymedphys/_dicom/dose.py [hunk 2]
```diff
@@ -179,58 +179,81 @@ def profile(displacements, depth, direction, dose_dataset, plan_dataset):
...
+        raise ValueError("Only one contour per slice is currently supported")
+
+    sorted_dose_z = np.sort(z_dose)
+
+    first_dose_index = np.where(sorted_dose_z == structure_z_values[0])[0][0]
+    for i, z_val in enumerate(structure_z_values):
+        dose_index = first_dose_index + i
+        if structure_z_values[i] != sorted_dose_z[dose_index]:
+            raise ValueError(
+                "Only contours where both, there are no gaps in the "
```
Comment (sjswerdloff) [inline pymedphys/_dicom/dose.py:212]: "Either there is a gap in the z-axis of the contours, or the contour axes and dose axes are not aligned"
and ideally, you'd want to have a different error message for the z-axis gap vs. axes misalignment.

File: pymedphys/_dicom/dose.py [hunk 3]
```diff
@@ -179,58 +179,81 @@ def profile(displacements, depth, direction, dose_dataset, plan_dataset):
     return extracted_dose
 
 
-def _get_indices(z_list, z_val):
-    indices = np.array([item[0] for item in z_list])
-    # This will error if more than one contour exists on a given slice
-    desired_indices = np.where(indices == z_val)[0]
-    # Multiple contour sets per slice not yet implemented
+def get_dose_grid_structure_mask(structure_name, structure_dataset, dose_dataset):
```
Comment (sjswerdloff) [inline pymedphys/_dicom/dose.py:182]: missing docstring 
which makes it harder to review

File: pymedphys/tests/dicom/structure/test_masking.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+        # All points outside the buffered mask are also outside the dicom_mask
+        assert np.all(
+            np.logical_and(
+                np.logical_not(buffered_mask), np.logical_not(dicom_mask[i, :, :])
+            )
+            == np.logical_not(buffered_mask)
+        )
+
+
+def get_grid_spacing(array):
```
Comment (sjswerdloff) [inline pymedphys/tests/dicom/structure/test_masking.py:77]: is this intended to be public?

File: pymedphys/tests/dicom/structure/test_masking.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    dx = np.unique(np.round(np.diff(array), 4))
+    assert len(dx) == 1
+
+    return dx
+
+
+def convert_contours_to_dummy_dicom_files(
```
Comment (sjswerdloff) [inline pymedphys/tests/dicom/structure/test_masking.py:84]: if it's intended to be public, there is enough going on here to include a docstring

File: pymedphys/tests/dicom/structure/test_masking.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    points_within_contour = shapely_points.intersection(with_buffer)
+
+    assert points_within_contour.within(with_buffer)
+
+    mask = slow_mask_with_loop(xx, yy, points_within_contour)
+
+    return mask
+
+
+def slow_mask_with_loop(xx, yy, points_within_contour):
```
Comment (sjswerdloff) [inline pymedphys/tests/dicom/structure/test_masking.py:160]: why is this named slow?

**Review by sjswerdloff — APPROVED** [review]
I would not hold back the PR, but please review my comments.

**Comment by SimonBiggs** [comment]
> I would not hold back the PR, but please review my comments.

Yup, I agree with all of those points. Thanks @sjswerdloff :)

**Comment by Matthew-Jennings** [comment]
It'd be nice to re-review this:
> Made the mask axis order be z, y, x to match `pymedphys.dicom.zyx_and_dose_from_dataset`

since I think we concluded in the past that:
- It'd be better to use exclusively use meshgrids when handling coords.
- Coord order should be xyz, not zyx
- zyx coord orders appear in the code as a quirk of preferring slice/row/column order when indexing/slicing either the dose grid or coord meshgrids. We should be careful not to confuse coords with indices, especially since indexing order should remain fixed as slice/row/column regardless of coord system (which can changes)

For a refresher on these points, see: #431

**Comment by SimonBiggs** [comment]
> It'd be nice to re-review this:
> 
> > Made the mask axis order be z, y, x to match `pymedphys.dicom.zyx_and_dose_from_dataset`
> 
> since I think we concluded in the past that:
> 
> * It'd be better to use exclusively use meshgrids when handling coords.
> * Coord order should be xyz, not zyx
> * zyx coord orders appear in the code as a quirk of preferring slice/row/column order when indexing/slicing either the dose grid or coord meshgrids. We should be careful not to confuse coords with indices, especially since indexing order should remain fixed as slice/row/column regardless of coord system (which can changes)
> 
> For a refresher on these points, see: #431

Yup, happy for you to mark this as "changes required". If you were interested in diving into the code itself that'd be awesome. For this one it has fallen way off of my priority list :/

**Comment by Matthew-Jennings** [comment]
> Yup, happy for you to mark this as "changes required". If you were interested in diving into the code itself that'd be awesome. For this one it has fallen way off of my priority list :/

If I dive in, would you like me to create a PR that merges into this one? Or an independent PR from `main`?

**Comment by SimonBiggs** [comment]
> > Yup, happy for you to mark this as "changes required". If you were interested in diving into the code itself that'd be awesome. For this one it has fallen way off of my priority list :/
> 
> If I dive in, would you like me to create a PR that merges into this one? Or an independent PR from `main`?

Up to you. It might be easier to mark this as draft and start a fresh if you'd prefer. Or if you'd like you can directly interact with this one.

It is important to me though that those notebook files continue to work, they were becoming quite nice "DICOM tutorial demos" that could be used as lecture material, workshop material or documentation how-tos

**Comment by SimonBiggs** [comment]
@Matthew-Jennings how about, I address @sjswerdloff comments here, merge this in, and then you can work on "adjusting it back"?

**Comment by Matthew-Jennings** [comment]
Sounds good 😊

**Comment by SimonBiggs** [comment]
@Matthew-Jennings, alright, she's all yours :).

**Pull request merged**